test/io/test_async_zarr.py

```diff
@@ -0,0 +1,590 @@
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "time",
+    [
+        [np.datetime64("1958-01-31")],
+        [
+            np.datetime64("1971-06-01T06:00:00"),
+            np.datetime64("2021-11-23T18:00:00"),
+            np.datetime64("2021-11-24T00:00:00"),
+        ],
+    ],
+)
+@pytest.mark.parametrize(
+    "variable",
+    [["t2m"], ["t2m", "tcwv"]],
+)
+@pytest.mark.parametrize(
+    "fs_factory",
+    [MemoryFileSystem, LocalFileSystem],
+)
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+async def test_async_zarr_write(
+    time: list[np.datetime64],
+    variable: list[str],
+    fs_factory: Callable[..., fsspec.spec.AbstractFileSystem],
+    device: str,
+    tmp_path: str,
+) -> None:
+
+    index_coords = {
+        "time": np.asarray(time),
+    }
+    z = AsyncZarrBackend(
+        f"{tmp_path}/output.zarr", index_coords=index_coords, fs_factory=fs_factory
+    )
+
+    total_coords = OrderedDict(
+        {
+            "time": np.asarray(time),
+            "variable": np.asarray(variable),
+            "lat": np.linspace(-90, 90, 180),
+            "lon": np.linspace(0, 360, 360, endpoint=False),
+        }
+    )
+    shape = [v.shape[0] for v in total_coords.values()]
+    x = torch.randn(shape, device=device, dtype=torch.float32)
+    for i, time0 in enumerate(time):
+        total_coords["time"] = np.array([time0])
+        z.write(x[i : i + 1], total_coords, "fields_1")
+        assert "fields_1" in [key async for key in z.root.array_keys()]
+        data = await (await z.root.get("fields_1")).getitem(slice(None))
+        assert data.shape == x.shape
+        assert np.allclose(data[i], x[i].to("cpu").numpy())
+    z.close()
+    data = await (await z.root.get("fields_1")).getitem(slice(None))
+    assert np.allclose(data, x.to("cpu").numpy())
+
+    total_coords = OrderedDict(
+        {
+            "variable": np.asarray(variable),
+            "time": np.asarray(time),
+            "lat": np.linspace(-90, 90, 180),
+            "lon": np.linspace(0, 360, 360, endpoint=False),
+        }
+    )
+    shape = [v.shape[0] for v in total_coords.values()]
+    x = torch.randn(shape, device=device, dtype=torch.float32)
+    for i, time0 in enumerate(time):
+        total_coords["time"] = np.array([time0])
+        z.write(x[:, i : i + 1], total_coords, "fields_2")
+        assert "fields_2" in [key async for key in z.root.array_keys()]
+        data = await (await z.root.get("fields_2")).getitem(slice(None))
+        assert data.shape == x.shape
+        assert np.allclose(data[:, i], x[:, i].to("cpu").numpy())
+    z.close()
+    data = await (await z.root.get("fields_2")).getitem(slice(None))
+    assert np.allclose(data, x.to("cpu").numpy())
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "fs_factory",
+    [MemoryFileSystem, LocalFileSystem],
+)
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+async def test_async_zarr_async_write(
```

Comment:
Seems mostly duplicated from the prev test, maybe could just parameterize whether or not to make the `write` async?